Objects/bytearrayobject.c

```diff
@@ -333,6 +333,42 @@ bytearray_iconcat_lock_held(PyObject *op, PyObject *other)
     _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(op);
     PyByteArrayObject *self = _PyByteArray_CAST(op);
 
+    // optimization: Avoid copying the bytes coming in when possible.
+    if (self->ob_alloc == 0 && _PyObject_IsUniquelyReferenced(other)) {
+        assert(_Py_IsImmortal(self->ob_bytes_object));
```

Comment:
I was thinking at something like:

```suggestion
        assert(self->ob_bytes_object == Py_GetConstantBorrow(Py_CONSTANT_EMPTY_BYTES));
```

Reply:
I'm worried what `Py_GetConstantBorrow` turns into in an optimized build as it's a C stable API funciton while `_Py_IsImmortal` is always inlined and I think provides the needed guarantee here (direct assigning to self->ob_bytes_object is safe).

Reply:
There's no “stable API” in PEP 652 :)
It's a *limited API* function, meaning that it compiles into stable ABI with the right settings. If you don't use the settings, it can be `static inline`. If it doesn't have a `static inline` implementation yet and you need it, add one. And if “add one” doesn't spark joy, please delegate to me; just ask.

But there's no point in chasing nanoseconds *here* -- in an optimized build, the whole `assert` disappears.

Reply:
(FWIW, doing it this way means *users* can handle the optimization/compatibility trade-offs with a recompile, without source changes.)

Reply:
I'm not sure if I understood your concern. In release mode, the assertion is removed.

Reply:
My assumptions around `assert` were wrong here; used to a codebase where `expr` was used regardless, but at least on my system `/usr/include/assert.h` + `NDEBUG` -> don't use expr at all.